### DISCUSSION TITLE
Malware reports

### DISCUSSION THREAD

**Discussion by hugogameiro** [mastodon/mastodon#18688] [General]
I received a report from my hosting provider saying that one of the servers I use to host Mastodon was being used as a botnet controller. So, they wanted me to clear the issue, or they could end up having to suspend my server.

I was puzzled because I was confident that the server was not infected, at least as sure as one can be after following security best practices. Still, I ran Rootkit Hunter and ClamAV to double-check, and nothing came up.

I could see on VirusTotal that the domain showed up with positive recent hits for what Virus Total calls 'Communicating Files' (files that communicate with this domain when opened or executed) and 'Files Referring' (files that contain the domain being studied in their bodies).
<img width="909" alt="Virus Total screenshot showing an example of a recent Communicating Files hit" src="https://user-images.githubusercontent.com/2203024/174430991-01481c0e-3f4b-4084-8312-3619dec73aeb.png">
<img width="911" alt="Virus Total screenshot showing an example of a recent Files Referring' hit" src="https://user-images.githubusercontent.com/2203024/174431043-06bb32da-33e9-41ba-9024-e71da2f29a22.png">

In fact, "6 security vendors flagged this domain as malicious", meaning that this domain was already being flagged as malware for some antivirus software.
<img width="1012" alt="Virus Total screenshot showing 6 security vendors marking the domain as malicious: Antiy-AVL, Certego, Dr.Web, Fortinet, Avira, CyRadar and Kaspersky" src="https://user-images.githubusercontent.com/2203024/174430759-e8e36f8a-4f16-4840-b868-679f43df5a62.png">

As I have little knowledge about these areas, I started investigating more. I found that Mastodon.social and Mastodon.online were also being flagged by 1 and 2 security vendors, respectively.
<img width="1018" alt="Virus Total screenshot for Mastodon.online showing 2 security vendors marking the domain as malicious: CRDF, Lumu" src="https://user-images.githubusercontent.com/2203024/174430922-290cf693-1d7a-4680-9e01-fd037cc6578d.png">

That's weird, right!? Mastodon.social and Mastodon.online are hosted by @Gargron on an entirely different network/infrastructure. Also, I’m pretty sure the Mastodon code is not infected, so I had to look for reasons this could be happening.

Then I came across [this post](https://zeltser.com/bots-command-and-control-via-social-media/) and similar papers [1](https://scholarworks.sjsu.edu/cgi/viewcontent.cgi?referer=&httpsredir=1&article=1246&context=etd_projects) [2](https://people.cs.umass.edu/~amir/papers/IH11-Stegobot.pdf), explaining how social media can be used for 'Command and Control' malware bots. 

In fact you can see that other social networks like [Twitter](https://www.virustotal.com/gui/domain/twitter.com/relations), [Facebook](https://www.virustotal.com/gui/domain/facebook.com/relations) or [Tumblr](https://www.virustotal.com/gui/domain/tumblr.com/relations) have also positive hits on VirusTotal for  'Communicating Files' and 'Files Referring'. The main difference is that security vendors do not mark these domains as malware. I guess they probably have them on some allowlist.

So, why are these reports happening if Mastodon code doesn't contain malware, and neither do those social media networks? From my basic understanding, what is happening is that an attacker creates a profile on a social media site like anyone else. Then adds the URL of their profile (or RSS feed) to the malware bot. Once the bot infects a third-party computer, it fetches the social media profile of the attacker and looks for instructions. The attacker can then issue instructions/commands using standard social media tools, and the bot follows those instructions. To give an illustration, the attacker could post instructing to 'take a screenshot and email it to evil  at address dot com' and even delete the post later, leaving no trace. This is not what I think usually what happens, but it's just something to give a basic idea on the process.

As you can imagine, this can somewhat explain why VirusTotal and similar projects that operate based on reports from automated detection mechanisms end up with many false positives. When one starts looking at the communications being made by an infected computer, the domain where the social media profile of the attacker is hosted starts to show up. So, the domain starts to be reported, leading to the social media website being marked as malware related.

On the domain in question, I found a suspicious profile @ronxik333, an empty account with only "Hello XXX.XXX.XXX.XXX|" (IP address) in the bio description. That profile was blocked. 

The next day, an account with the username @olegf9844n was created with the same "Hello XXX.XXX.XXX.XXX|" but a different IP. That account has also been deleted. Here is a screenshot of the bio:
<img width="422" alt="Screenshot of the profile bio with the Hello IP address pattern" src="https://user-images.githubusercontent.com/2203024/174432536-d7df28d1-0f66-4bd7-a03d-afb4b7b9a409.png">

Yesterday, I realised that @ronxik333 also exists on Mastodon.online. 

I believe that something like this is what is happening and causing the malware reports on both the domain I host, Mastodon.social, Mastodon.online and probably others that I have not looked into.

As another confirmation of my theory searching the domains on [CRDF Threat Center](https://threatcenter.crdf.fr/forensic_domains.html)  ended up showing up the @ronxik333 profile on both domains as the only URL being reported as a 'Malicious:URL'.
<img width="578" alt="CRDF Threat Center screenshot showing the URL mastodon.online/@ronxik13 as Malicious:URL" src="https://user-images.githubusercontent.com/2203024/174433119-7225da51-4c2f-42c7-b023-d765315c019b.png">

Sorry for the long post, but I wanted to share this with other Mastodon admins. I think this is something to be on the lookout for. In particular, if you run a Mastodon instance that has open registrations. I'm not sure there is much we can do to mitigate this besides looking out and solving problems when they arise, but if someone has thoughts or suggestions, please share them.

**Comment by nightpool** [comment]
Wow, thanks for the deep dive! It might be helpful if you also posted the registration email address and IP for @ronxik133 so that admins can block them more widely or we can see if they have other usernames elsewhere.

**Comment by hugogameiro** [comment]
@nightpool I found 2 accounts, @ronxik333 and @olegf9844n. Both were registered with gmail.com addresses and had the same handle as the usernames created. 

The IPs, I found had no pattern that I could detect. The account creation and the account confirmation/first login, they were all different, different countries. One was from a regular ISP and the others from different data center. 

Also, the IP used on the "Hello XXX.XXX.XXX.XXX|" pattern from both accounts was different.

I think it will be hard to block based on IPs or email. What I did was search in the database for an IP in the bio using:
`select id, username, note from accounts where domain is null and note ~ '[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}';`

And even (because I am paranoid) in the statuses (but this one gives a lot of false positives):
`select statuses.id, username, text from statuses, accounts where accounts.id=account_id and domain is null and text ~ '[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}';`

Still, if someone wants the IPs or more information, please DM me and I will gladly share it.

**Reply by lfuelling** [reply]
The IP in the screenshot seems to be hosted by Hetzner. In my experience they are trustworthy when it comes to handling abuse reports; You should report the IPs (that have a WHOIS pointing to Hetzner) to `abuse[at]hetzner[dot]com`.

**Comment by SilverBut** [comment]
This is a pretty usual way (by "this" I mean using legal services to distribute malware's Command and Control, or C2 server address). 

For site owner, find our the problematic user is necessary. In different cases, the malware author may decided to distribute the address in different ways:

- Sending plain text. Like your case. It might be in toots, intros, or somewhere else. Those users will weird enough. This may also happen in replies. For example, one bot account send some sentences generated by GPT3 or copied from hackernews. The other bot, besides sending toots, would also always reply the first account's toot with plain C2 server address. 
- Send some obfuscated/encrypted text. This is also possible to be noticed, since they also "seems weird".
- Send carefully encoded text. For example, send latest solidot news per 15 minute, but modify amounts of spaces, so you can represent an IPv4 address in an hour. This is hard to find. Those users looks harmless, sometimes attacker is the real human user under those accounts, and communicate with other server members to increase credit. 

And this is just about how to _send_ out the message. The receiver may have some different strategy:

- Crawl some user's first toot by using webpage with/without creds
- Get server site-local trending toot list
- Search on server

Note I'm just listing possible ways to abuse Mastodon. Those ways have been observed in other social media platforms, and may/may not exists on the Fediverse. 

Here are some personal suggestions.

For site owner, if you got a warn from ISP about this, try locate the user, submit the user info (if law permits) and ask if the user should be banned. Usually they would ask you to do so, so you can prove you are not related with the event.

If the warn comes from court order or something like this, I suggest you locate the user first, and talk with your lawyer. If you decided to follow the order, follow it. Sometimes the police would just ask you not banning the account until they can take control of the full zombie network. 

If you have difficulties in locating the user, ISP should provide some additional information like which URL/requests may related with the malware. In your case, you may also want to see if VirusTotal can provide information about why they linked your site with the malware. 

Preventing such an "attack" is hard. Server to server communicate depends on HTTP so you can not just add Cloudflare in the front and enable highest defense mode. This would harm the federation. 

Consider enable invite-only mode or review every new user application. Regularly check your access log to see if any weird behavior patterns, for example, unusual user agent, repeatedly visit one url from lots of different addresses. 

I'm wondering if it is possible to "subscribe" a block list including known email address and IPs. So the community can share those threat intelligence and may link with other TI providers.

**Comment by michft** [comment]
So to improve security I would;
1 make sure there is a standard way to contact the Admin of a server. A .well_known/security.txt contacts file. https://securitytxt.org/ or a dns security txt record https://github.com/disclose/dnssecuritytxt/tree/main/docs.  This allows people who control servers to get contacted so they can act on problems. Do this on top of Admin contact details for site Users 
2 Run as many OWASP tools on the standard-site as feasible (there is always cost/benefit return). https://owasp.org/ this will stop people getting Admin when they shouldn't, or accessing stuff they shouldn't etc
2b Definitely have a shared way for Mastodon Admins to do updates and be informed when someone tries to attack one server (if they succeed on one they will try more!)
3 Run as many Malware scanners as feasible on your own server
4 You have Github Dependabot library-checking turned on. YEAH!!
5 Maybe do some URLHaus or similar URL known Malware site checking and proactive blocking. https://urlhaus.abuse.ch/ Basically a blacklist of postable URL's (https://developers.google.com/safe-browsing/ as well, main problem is usual API tracking, URL Harvesting)
6 Maybe server policy to allow for no links or URLs (probably too draconian)

**Reply by nightpool** [reply]
How would any of these things prevent users from signing up for an account and using it to send coded messages to spambots? This post is entirely off topic for this thread

**Comment by shleeable** [comment]
I've recently installed crowdsec to test that out. It should hopefully block some malware from accessing the server. Hopefully.

This is a tool that seems to keep a large regularly updated blacklist that it maintains in iptables.
```
Name: crowdsec-blacklists
Type: hash:net
Revision: 6
Header: family inet hashsize 8192 maxelem 65536 timeout 300
Size in memory: 687608
References: 1
Number of entries: 14786
Members:
157.230.113.181 timeout 598292
96.241.40.16 timeout 414690
etc
```

**Comment by wiligl** [comment]
I would urge key developers in the Mastodon community to solve this issue soon because after the takeover of Twitter by Elon Musk, there is a "golden opportunity" to grow the Mastodon community with unhappy Twitter users. However, virus scanner reporting "malicious" Mastodon websites will scare many potential Mastodon users off. At a minimum some guideline should be published, ideally by Eugen Rochko, on the size of the actual safety issue and how to handle it.
